### PR TITLE
feat: Payment Split

### LINKED ISSUE #16: [TASK] Add `operatorAddressFromPublicKey` to the BlueprintServiceManager Base contract
```sol
    /**
     * @dev Converts a public key to an operator address.
     * @param publicKey The public key to convert.
     * @return operator address The operator address.
     */
    function operatorAddressFromPublicKey(bytes calldata publicKey) internal pure returns (address operator) {
        return address(uint160(uint256(keccak256(publicKey))));
    }
```

### LINKED ISSUE #24: [TASK] Implement payment split mechanism
```frontmatter
- **Priority**: P0
```

### Description
Implement the core payment split mechanism that distributes payments according to the configurable ratios.

Here is an example:
- Developers: 50%
- Protocol: 20%
- Operators/Restakers: 30% (Operators 10%, Restakers 20%)

### Expected Outcomes
- Smart contract implementing payment splits
- Accurate distribution calculations
- Gas-efficient distribution mechanism

### PR DESCRIPTION
**Pull request by shekohex**
This pull request includes significant changes to the asset management system in the blockchain project. The main focus is the introduction of the `AssetsLib.sol` library and the refactoring of the `BlueprintServiceManagerBase` contract to utilize this new library. Additionally, some redundant code has been removed, and new utility functions have been added.

### Introduction of `AssetsLib.sol` library:
* Added a new library `AssetsLib.sol` to handle asset types and conversions, including enums, structs, and utility functions for asset management.

### Refactoring to use `AssetsLib.sol`:
* Updated `BlueprintServiceManagerBase` to use the new `Assets` library for asset management, replacing the old `ServiceOperators` library. This includes changes to functions such as `_permitAsset`, `_revokeAsset`, `_getPermittedAssets`, and `_isAssetPermitted`. [[1]](diffhunk://#diff-6fe79b83efb9fdfb7f295ea335ff1232519b7ed772eec108195efd47df7bef91R19-R21) [[2]](diffhunk://#diff-6fe79b83efb9fdfb7f295ea335ff1232519b7ed772eec108195efd47df7bef91L31-L33) [[3]](diffhunk://#diff-6fe79b83efb9fdfb7f295ea335ff1232519b7ed772eec108195efd47df7bef91L179-R179) [[4]](diffhunk://#diff-6fe79b83efb9fdfb7f295ea335ff1232519b7ed772eec108195efd47df7bef91L195-L213) [[5]](diffhunk://#diff-6fe79b83efb9fdfb7f295ea335ff1232519b7ed772eec108195efd47df7bef91L222-L240) [[6]](diffhunk://#diff-6fe79b83efb9fdfb7f295ea335ff1232519b7ed772eec108195efd47df7bef91L282-R260) [[7]](diffhunk://#diff-6fe79b83efb9fdfb7f295ea335ff1232519b7ed772eec108195efd47df7bef91L311-R278)

### Removal of redundant code:
* Removed the `ServiceOperators` library from `IBlueprintServiceManager.sol` and replaced it with an import statement for the new `ServiceOperatorsLib.sol`. [[1]](diffhunk://#diff-d5a72da0582dfb5e9d96eddf40c940f7a3d90018136cfefd3139bb388dde2ec4L4-R4) [[2]](diffhunk://#diff-d5a72da0582dfb5e9d96eddf40c940f7a3d90018136cfefd3139bb388dde2ec4L234-R169)

### Additional imports and utility usage:
* Added new imports and utility functions in `MasterBlueprintServiceManager.sol` to support the `Assets` library and `SafeERC20` for handling ERC20 tokens. [[1]](diffhunk://#diff-0c5d69e5401ffb3c6c8fd5b531680e8e6e165ee4ba6a9bf9ea3dcb8ee1de5d16R7-R8) [[2]](diffhunk://#diff-0c5d69e5401ffb3c6c8fd5b531680e8e6e165ee4ba6a9bf9ea3dcb8ee1de5d16R18-R31)

These changes improve the modularity and readability of the codebase by centralizing asset management logic within a dedicated library.

Closes #24 
Closes #16

### CODE REVIEW

File: src/MasterBlueprintServiceManager.sol [hunk 1]
```diff
@@ -454,4 +470,44 @@ contract MasterBlueprintServiceManager is RootChainEnabled, AccessControl, Pausa
...
+    )
+        internal
+    {
+        // TODO: make the following logic dynamic and configurable.
+        // Here is an example:
+        // - Developers: 50%
+        // - Protocol: 20%
+        // - Operators/Restakers: 30% (Operators 10%, Restakers 20%)
+        uint256 totalAmount = request.amount;
+        uint256 developerAmount = (totalAmount * 50) / 100;
```
Comment (drewstone) [inline src/MasterBlueprintServiceManager.sol:487]: Parameters should come from onchain with values we can update

File: src/MasterBlueprintServiceManager.sol [hunk 2]
```diff
@@ -39,6 +55,34 @@ contract MasterBlueprintServiceManager is RootChainEnabled, AccessControl, Pausa
...
+        /// @notice Blueprint Restakers trench
+        /// @dev The restakers trench is for the restakers of the blueprint.
+        Restakers
+    }
+
+    /// @dev Defines the trench with the percentage.
+    /// @notice Use this struct to define the trench with the percentage.
+    struct Trench {
+        /// @dev The kind of the trench.
+        TrenchKind kind;
```
Comment (drewstone) [inline src/MasterBlueprintServiceManager.sol:79]: TrancheKind

File: src/MasterBlueprintServiceManager.sol [hunk 3]
```diff
@@ -39,6 +55,34 @@ contract MasterBlueprintServiceManager is RootChainEnabled, AccessControl, Pausa
...
+        /// @dev The operators trench is for the operators of the blueprint.
+        Operators,
+        /// @notice Blueprint Restakers trench
+        /// @dev The restakers trench is for the restakers of the blueprint.
+        Restakers
+    }
+
+    /// @dev Defines the trench with the percentage.
+    /// @notice Use this struct to define the trench with the percentage.
+    struct Trench {
```
[the remark this answers is not in the document]
Comment (shekohex) [reply src/MasterBlueprintServiceManager.sol:77]: Thanks, these typos will be fixed shortly.

File: src/MasterBlueprintServiceManager.sol [hunk 4]
```diff
@@ -454,4 +535,108 @@ contract MasterBlueprintServiceManager is RootChainEnabled, AccessControl, Pausa
...
+    /// @param _trenches The array of trenches and their percentages.
+    /// @notice Only the trench updater can update the trenches.
+    function setTrenches(Trench[] calldata _trenches) public onlyRole(TRENCH_UPDATE_ROLE) {
+        _setTrenches(_trenches);
+    }
+
+    /// @dev Initialize the default trenches for the blueprint.
+    function _intializeDefaultTrenches() internal {
+        Trench[] memory _trenches = new Trench[](4);
+        _trenches[0] = Trench(TrenchKind.Developer, 5000); // 50%
```
Comment (drewstone) [inline src/MasterBlueprintServiceManager.sol:561]: Shouldn't these % come from the pallet?

**Pull request merged**